pennylane/ops/functions/matrix.py

```diff
@@ -134,13 +134,27 @@ def _matrix(tape, wire_order=None):
     interface = qml.math._multi_dispatch(params)
 
     wire_order = wire_order or tape.wires
-    n_wires = len(wire_order)
 
     # initialize the unitary matrix
-    unitary_matrix = qml.math.eye(2**n_wires, like=interface)
+    unitary_matrix = qml.math.eye(2 ** len(wire_order), like=interface)
 
+    broadcasted_unitary = False
     for op in tape.operations:
         U = matrix(op, wire_order=wire_order)
-        unitary_matrix = qml.math.tensordot(U, unitary_matrix, axes=[[-1], [-2]])
+        broadcasted_U = qml.math.ndim(U) == 3
+        if broadcasted_U and broadcasted_unitary:
+            # If both, U and unitary_matrix are broadcasted, we need a special syntax
+            unitary_matrix = qml.math.stack(
+                [qml.math.dot(u, _unitary) for u, _unitary in zip(U, unitary_matrix)]
+            )
+        else:
+            # This covers the cases where at most one of U and unitary_matrix is broadcasted
+            unitary_matrix = qml.math.tensordot(U, unitary_matrix, axes=[[-1], [-2]])
+            # If unitary_matrix was broadcasted, we need to move the corresponding axis up front
+            if broadcasted_unitary:
+                unitary_matrix = qml.math.moveaxis(unitary_matrix, 1, 0)
+
+        if broadcasted_U:

```

Comment:
Likely okay to set this once only.
```suggestion
        if not broadcasted_unitary and broadcasted_U:
```

Reply:
Time-wise the difference (as well as the overall consumed time) is really negligible, so we're down to readability. Any preference?

Reply:
Hmm, I'd imagine the suggested version to be faster just because Python would evaluate the expression lazily (i.e., once `not broadcasted_unitary is False` `and broadcasted_U` is not being evaluated).

Readabilitywise, imho using improved names would make this block readable too.

Reply:
I agree, but because the loop only contains the assignment of a Boolean, it is similarly cheap as checking a Boolean, I think. In any case, this is about nanoseconds...

I updated the names, feel free to choose whatever is best here and I'll include it :)

Reply:
This is outdated now :)